### PR TITLE
Migrate from plain manifests to helm charts in mesheryctl

### LINKED ISSUE #4103: [mesheryctl] Migrate from Manifests to Helm for Meshery Server/Adapter deployment
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior
1. Server and Adapters are provisioned with manifests.
1. Meshery Operator are  provisioned with helm charts, which uses `ServiceAccountName` and that dependencies are provisioned in correct order of operation. 

#### Desired Behavior
1. Use MeshKit helm chart deployment functionality using https://meshery.io/charts.

---
#### Contributor Resources
- [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by panyuenlau**
**Description**

This PR fixes the helm chart part of #4103
(The `mesheryctl` changes will be included in another PR to separate from the helm chart changes)

**Notes for Reviewers**
The current PR only fixes the existing issues related to `serviceAccount`, `clusterRole`, and `clsuterRoleBInding` within the helm charts, I've tested the basic installations for each adapter but need some help on testing the other functionalities

> Note: all the adapters should just use the service account `meshery-server` to have the permission to perform certain operations, but some of the adapters had their own service accounts created in the helm charts

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

~>TODO: will be adding changes to the `mesheryctl` implementation to use the helm chart after making sure the charts are functioning properly~

### CODE REVIEW

File: install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml [hunk 1]
```diff
@@ -30,7 +30,14 @@ rules:
   - patch
   - update
   - watch
-  - approve
+- apiGroups:
+  - meshery.layer5.io
+  resources:
+  - brokers/status
+  verbs:
```
Comment (Aisuko) [inline install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml:37]: Please give any code comment for apiGroups' `verbs` if could what I mean we should sure which permissions we need to get.

**Review by Aisuko — CHANGES_REQUESTED** [review]
Thanks, @panyuenlau, so cleanly work. Please add the code comment to the  `ClusterRole` in order to clearly describe the permission we need to apply.

File: install/kubernetes/helm/meshery/charts/meshery-consul/values.yaml [hunk 2]
```diff
@@ -26,10 +26,10 @@ serviceAccount:
   # we need to split to every adapter with themself's cluster role. Although there a lot of work to do but can make the charts flexible.
 
   # Specifies whether a service account should be created
-  create: true
+  create: false
   # The name of the service account to use.
   # If not set and create is true, a name is generated using the fullname template
-  name: meshery-adapter-consul
+  name: meshery-server
```
Comment (Aisuko) [inline install/kubernetes/helm/meshery/charts/meshery-consul/values.yaml:32]: I have a consider here, please take a look on this 

https://github.com/meshery/meshery/blob/df44458404f7456b8d5e15434dcce13e912f06f2/install/kubernetes/helm/meshery/values.yaml#L34-L37

If the end-user replaces the default value of `service-account` from `meshery-server` to others that may cause the adapters can not work. right?

So, here we have a solution many people already use that is we set up the value on the top, we can set the value from the top chart to replace the `sub-charts` specific value. That can be good.

As you can see below, we can replace `fullnameOverride` for every `sub-charts`, and we can do the same for the `sa`.

https://github.com/meshery/meshery/blob/df44458404f7456b8d5e15434dcce13e912f06f2/install/kubernetes/helm/meshery/values.yaml#L93-L130


So, you can do it in the next PR, but please add the new relative issue for this.

**Comment by Aisuko** [comment]
Hi, @navendu-pottekkat, I remember the CI check was only triggered by the specific folder, right? Is that strategy keep working well? I saw there trigger so many CI includes UI, DOC, `golangci-lint`. please help us check the strategy if possible.

File: install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml [hunk 3]
```diff
@@ -11,10 +11,10 @@ image:
 env: {}
 
 probe:
-  #TODO: Need to confirm the health check path of meshery. 
+  #TODO: Need to confirm the health check path of meshery.
```
Comment (s1ntaxe770r) [inline install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml:14]: @navendu-pottekkat  what's Meshery's health check path?

File: install/kubernetes/helm/meshery/charts/meshery-consul/values.yaml (hunk 2 above)
Comment (hexxdump) [reply install/kubernetes/helm/meshery/charts/meshery-consul/values.yaml:32]: > I have a consider here, please take a look on this
> 
> https://github.com/meshery/meshery/blob/df44458404f7456b8d5e15434dcce13e912f06f2/install/kubernetes/helm/meshery/values.yaml#L34-L37
> 
> If the end-user replaces the default value of `service-account` from `meshery-server` to others that may cause the adapters can not work. right?
> 
> So, here we have a solution many people already use that is we set up the value on the top, we can set the value from the top chart to replace the `sub-charts` specific value. That can be good.
> 
> As you can see below, we can replace `fullnameOverride` for every `sub-charts`, and we can do the same for the `sa`.
> 
> https://github.com/meshery/meshery/blob/df44458404f7456b8d5e15434dcce13e912f06f2/install/kubernetes/helm/meshery/values.yaml#L93-L130
> 
> So, you can do it in the next PR, but please add the new relative issue for this.

+1 to what @Aisuko suggested

File: install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml (hunk 1 above)
Comment (hexxdump) [reply install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml:37]: Also, should the templates/NOTES.txt be modified to capture the recent changes?

**Comment by hexxdump** [comment]
> **Description**
> 
> This PR fixes #4103
> 
> **Notes for Reviewers** The current PR only fixes the existing issues related to `serviceAccount`, `clusterRole`, and `clsuterRoleBInding` within the helm charts, I've tested the basic installations for each adapter but need some help on testing the other functionalities
> 
> > Note: all the adapters should just use the service account `meshery-server` to have the permission to perform certain operations, but some of the adapters had their own service accounts created in the helm charts
> 
> **[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
> 
> * [x]  Yes, I signed my commits.
> 
> > TODO: will be adding changes to the `mesheryctl` implementation to use the helm chart after making sure the charts are functioning properly

I feel, the mesheryctl changes(to move away from manifests) should be handled through a separate PR

**Comment by panyuenlau** [comment]
> > **Description**
> > This PR fixes #4103
> > **Notes for Reviewers** The current PR only fixes the existing issues related to `serviceAccount`, `clusterRole`, and `clsuterRoleBInding` within the helm charts, I've tested the basic installations for each adapter but need some help on testing the other functionalities
> > > Note: all the adapters should just use the service account `meshery-server` to have the permission to perform certain operations, but some of the adapters had their own service accounts created in the helm charts
> > 
> > 
> > **[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
> > 
> > * [x]   Yes, I signed my commits.
> > 
> > > TODO: will be adding changes to the `mesheryctl` implementation to use the helm chart after making sure the charts are functioning properly
> 
> I feel, the mesheryctl changes(to move away from manifests) should be handled through a separate PR

I agree, since separating the two into different PRs would help us identify potential issues more easily

File: install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml (hunk 3 above)
Comment (piyushsingariya) [reply install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml:14]: There is no specific health-check path but we can try hitting the version endpoint to get the surety that meshery is running properly.

File: install/kubernetes/helm/meshery/values.yaml [hunk 4]
```diff
@@ -117,6 +117,10 @@ meshery-nsm:
   enabled: true
   fullnameOverride: meshery-nsm
 
+meshery-nginx-sm:
+  enabled: true
+  fullnameOverride: meshery-nginx-sm
+
```
[the remark this answers is not in the document]
Comment (panyuenlau) [reply install/kubernetes/helm/meshery/values.yaml:129]: @hexxdump I believe we do have the meshery-nsm helm chart: https://github.com/meshery/meshery/tree/master/install/kubernetes/helm/meshery/charts/meshery-nsm

File: install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml (hunk 1 above)
Comment (panyuenlau) [reply install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml:37]: @Aisuko @hexxdump TBH, I am not familiar enough with the system to give a good explanation on why we need those permissions for the cluster role, I made the changes just based on the plain yaml manifests that we're currently using... reference: https://github.com/meshery/meshery-operator/blob/master/config/manifests/default.yaml#L273-L288

File: install/kubernetes/helm/meshery/values.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (leecalcote) [reply install/kubernetes/helm/meshery/values.yaml:129]: Good question and answer. The other two adapters are alpha stage adapters, and so, charts haven't been created yet. We can go ahead and create those charts now, so long as we don't include their deployment by default.

File: install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml (hunk 3 above)
Comment (leecalcote) [reply install/kubernetes/helm/meshery/charts/meshery-cpx/values.yaml:14]: A good item to create a new issue for follow up here.

File: install/kubernetes/helm/meshery/charts/meshery-nginx-sm/values.yaml [hunk 5]
```diff
@@ -49,7 +52,7 @@ securityContext: {}
 service:
   type: ClusterIP
   port: 10010
-  targetPOrt: 10010
```
Comment (leecalcote) [inline install/kubernetes/helm/meshery/charts/meshery-nginx-sm/values.yaml:52]: Yikes!

File: install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml (hunk 1 above)
Comment (leecalcote) [reply install/kubernetes/helm/meshery/charts/meshery-operator/templates/clusterrole.yaml:37]: Good discussion. Given that the scope of these verbs is restricted to `meshery.layer5.io` resource groups, we shouldn't see any pushback from users given that the Meshery Operator should have full control over the lifecycle of its custom controllers.

**Review by leecalcote — COMMENTED** [review]
@panyuenlau confirming: with this update, two service accounts will be created: `meshery-server` and `meshery-operator`?

**Comment by panyuenlau** [comment]
> @panyuenlau confirming: with this update, two service accounts will be created: `meshery-server` and `meshery-operator`?

Correct, the `meshery` and all the adapters should share the `meshery-server` service account in order to use the UI to manage the corresponding service mesh; and `meshery-operator` service account is for the `meshery-operator` pod

**Pull request merged**